controllers/machinehealthcheck_targets.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	machinePhaseFailed = "Failed"
+
+	// Timeout waiting for the provider to bring up a node
+	timeoutForMachineToHaveNode = 10 * time.Minute
+
+	// EventDetectedUnhealthy is emitted in case a node associated with a
+	// machine was detected unhealthy
+	EventDetectedUnhealthy string = "DetectedUnhealthy"
+)
+
+// healthCheckTarget contains the information required to perform a health check
+// on the node to determine if any remediation is required.
+type healthCheckTarget struct {
+	Machine *clusterv1.Machine
+	Node    *corev1.Node
+	MHC     *clusterv1.MachineHealthCheck
+}
+
+func (t *healthCheckTarget) string() string {
+	return fmt.Sprintf("%s/%s/%s/%s",
+		t.MHC.GetNamespace(),
+		t.MHC.GetName(),
+		t.Machine.GetName(),
+		t.nodeName(),
+	)
+}
+
+// Get the node name if the target has a node
+func (t *healthCheckTarget) nodeName() string {
+	if t.Node != nil {
+		return t.Node.GetName()
+	}
+	return ""
+}
+
+// Determine whether or not a given target needs remediation
+func (t *healthCheckTarget) needsRemediation(logger logr.Logger) (bool, time.Duration) {
+	var nextCheckTimes []time.Duration
+	now := time.Now()
+
+	// machine has failed
+	if t.Machine.Status.Phase == machinePhaseFailed {
```

Comment:
Although this seems ok to me, there was discussion in the proposal where the majority preferred not to reference a phase value directly since they are meant to be informative.

Reply:
Should this check for `t.Machine.Status.FailureReason != nil` instead?

Reply:
I think that works and has the same effect, have changed it